engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/number/bound/decimal/DecimalNumberComparatorHelper.java

```diff
@@ -32,8 +32,26 @@ public static int compare(Long number, BigDecimal value) {
 		return BigDecimal.valueOf( number ).compareTo( value );
 	}
 
-	public static int compare(Number number, BigDecimal value) {
-		return BigDecimal.valueOf( number.doubleValue() ).compareTo( value );
+	public static int compare(Number number, BigDecimal value, OptionalInt treatNanAs) {
+		// In case of comparing numbers before we compare them as two longs:
+		// 1. We need to check for floating point number as it should be treated differently in such case:
+		if ( number instanceof Double ) {
+			return compare( (Double) number, value, treatNanAs );
+		}
+		if ( number instanceof Float ) {
+			return compare( (Float) number, value, treatNanAs );
+		}
+
+		// 2. We need to check for big numbers so that we don't lose data when converting them to long:
+		if ( number instanceof BigDecimal ) {
+			return compare( (BigDecimal) number, value );
+		}
+		if ( number instanceof BigInteger ) {
+			return compare( (BigInteger) number, value );
+		}
+
+		// if we reached this step only integer numbers remains hence we convert to long:
+		return BigDecimal.valueOf( number.longValue() ).compareTo( value );
```

Comment:
Hmmm. I'm not sure it's such a good idea. People could have their specific decimal implementation of `Number`.

I think you could have special cases but at the end you should use `doubleValue()`. BTW, if we do that, all the comments of the method needs updating.

Reply:
Honestly if someone have their own number I would expect them to add their own validator :smiley: 
Thinking more about it I think we should do 
```java
new BigDecimal( number.toString() ).compareTo( value )
```
And then there's the same situation in `NumberComparatorHelper` where we convert to long. We should do the same `BigDecimal` check there as well then. WDYT ?

Reply:
I don't think using `toString()` will be reliable enough. What I would do is special case all the known types and end up with a `double` based comparison for both `Max/Min` and `DecimalMax/Min`.

Does it make sense to you?

Reply:
yes, I'll update the PR soon